server/src/main/java/org/elasticsearch/indices/SystemIndices.java

```diff
@@ -317,6 +330,19 @@ private static Automaton featureToIndexAutomaton(Feature feature) {
         return systemIndexAutomaton.orElse(EMPTY);
     }
 
+    private static Automaton buildDataStreamAutomaton(Map<String, Feature> descriptors) {
```

Comment:
I agree, I'd prefer to not have two different implementations like this. If having them separate is necessary we should clarify why and how they're different.

The only obvious difference to me is that the `Predicate` version does not support patterns - i.e., that predicate will only return `true` if there's an exact match of the data stream name. I'm not sure what, if anything, is using a pattern for the data stream name.

Reply:
Since both implementations were built from the same Java stream of system data stream names, I've reimplemented the predicate on top of the automaton.